drivers/overlay/overlay.go

```diff
@@ -264,23 +264,39 @@ func parseOptions(options []string) (*overlayOptions, error) {
 	return o, nil
 }
 
-func supportsOverlay() error {
+func supportsOverlay(home string, rootUID, rootGID int) error {
 	// We can try to modprobe overlay first before looking at
 	// proc/filesystems for when overlay is supported
 	exec.Command("modprobe", "overlay").Run()
 
-	f, err := os.Open("/proc/filesystems")
-	if err != nil {
-		return err
-	}
-	defer f.Close()
-
-	s := bufio.NewScanner(f)
-	for s.Scan() {
-		if s.Text() == "nodev\toverlay" {
-			return nil
+	// Try a test mount in the specific location we're looking at using.
+	layerDir, err := ioutil.TempDir(home, "compat")
+	if err == nil {
+		mergedDir := filepath.Join(layerDir, "merged")
+		upperDir := filepath.Join(layerDir, "diff")
+		workDir := filepath.Join(layerDir, "work")
+		lowerDir := filepath.Join(layerDir, "lower")
+		defer func() {
+			// Permitted to fail, since the various subdirectories
+			// can be empty or not even there, and the home might
+			// legitimately be not empty
+			_ = unix.Unmount(mergedDir, unix.MNT_DETACH)
+			_ = os.RemoveAll(layerDir)
+			_ = os.Remove(home)
+		}()
+		_ = idtools.MkdirAs(mergedDir, 0700, rootUID, rootGID)
```

Comment:
@rhvgoyal Can we get away with fewer directories.  Do we actually need an upper directory?

Reply:
If we wanted to mount the overlay for read-only use, I think we could.  The overlay filesystem has requirements for the upper directory that it doesn't have for lower directories, though, so I think it's more useful to mimic the eventual use case, to avoid false "yeah, it's fine" results.
This function call was moved after the checks on the kernel version and of the underlying filesystem's type, so that when we'll reject the driver based on one of those checks, we won't bother with this check.

Reply:
Upper dir is not must. If upper is not being passed in, then workdir is not required as well. All we need is multiple lower dirs and one merged dir.

Reply:
In the general case you're right, but all of the mounts we end up making in the driver use an upper, and I think the kernel's check on whether or not the upper directory supports d_type is a useful thing to notice at this point, so I'm inclined to leave that detail unchanged unless someone insists that it be changed.

Reply:
Kernel d_type check just emits  a warning. So it does not give you any extra information whether upper supports d_type or not. I think that will have to be a separate test. If filesystem is xfs, run xfs_info and check ftype=1.

Reply:
Hmm, that's a good point: if the kernel's not going to reject that mount attempt, we learn nothing.  Moving the existing d_type check to be here so that it can use the temporary directory that we're already creating.